js/src/ui/Form/AddressSelect/addressSelect.js

```diff
@@ -15,262 +15,799 @@
     }
 
-    return this.items[address];
+    return (
+      <div
+
+      >
+        { this.renderAccountCard({ address: inputValue }) }
+      </div>
+    );
   }
 
-  getBalance (address) {
-    const { balances = {} } = this.props;
-    const balance = balances[address];
+  renderCloseIcon () {
+    const { expanded } = this.state;
 
-    if (!balance) {
+    if (!expanded) {
       return null;
     }
 
-    const ethToken = balance.tokens.find((tok) => tok.token && tok.token.tag && tok.token.tag.toLowerCase() === 'eth');
+    return (
+      <div className={ styles.closeIcon } onClick={ this.handleClose }>
+        <CloseIcon style={ { width: 48, height: 48 } } />
+      </div>
+    );
+  }
 
-    if (!ethToken) {
-      return null;
+  renderAccounts () {
+    const { values } = this.state;
+
+    if (values.length === 0) {
+      return (
+        <div className={ styles.categories }>
+          <div className={ styles.empty }>
+            No account matches this query...
+          </div>
+        </div>
+      );
     }
 
-    return ethToken.value;
+    const categories = values.map((category) => {
+      return this.renderCategory(category.label, category.values);
+    });
+
+    return (
+      <div className={ styles.categories }>
+        { categories }
+      </div>
+    );
   }
 
-  renderBalance (address) {
-    const balance = this.getBalance(address);
-    const value = fromWei(balance);
+  renderCategory (name, values = []) {
+    if (values.length === 0) {
+      return null;
+    }
+
+    const cards = values
+      .map((account) => this.renderAccountCard(account));
 
     return (
-      <div className={ styles.balance }>
-        { value.toFormat(3) }<small> { 'ETH' }</small>
+      <div className={ styles.category } key={ name }>
+        <div className={ styles.title }>{ name }</div>
+        <div className={ styles.cards }>
+          <div>{ cards }</div>
+        </div>
       </div>
     );
   }
 
-  renderMenuItem (address) {
-    const balance = this.props.balances
-      ? this.renderBalance(address)
+  renderAccountCard (_account) {
+    const { copied } = this.state;
+    const { address, index = null } = _account;
+
+    const account = this.props.accountsInfo[address];
+    const name = (account && account.name && account.name.toUpperCase()) || address;
+    const { tags = [] } = account && account.meta || {};
+
+    const balance = this.renderBalance(address);
+
+    const onClick = () => {
+      this.handleClick(address);
+    };
+
+    const onFocus = () => {
+      this.setState({ focusedItem: index });
+    };
+
+    const classes = [ styles.account ];
+
+    if (index && copied === index) {
+      classes.push(styles.copied);
+    }
+
+    const addressElements = name !== address
+      ? (
+        <div className={ styles.addressContainer }>
+          <span
+            className={ styles.address }
+            onClick={ this.preventEvent }
+            ref={ `address_${index}` }
```

Comment:
We use a lot of references, which means a lot of manipulation - consider keeping it light. (Here and elsewhere)